### PR TITLE
Adopt metals implementation for symbol inspection

### LINKED ISSUE #183: Shift-tab tooltip for scaladoc/javadoc/parameters hint
Python kernel shows docstring tooltip on shift-tab, which is very useful for finding information on function and parameters:
tooltip

It is possible to support something like this for scala kernel? 

If somebody can point me in the right direction, I can try to implement it myself.

Thanks.

### LINKED ISSUE #764: Error starting metabrowse server
Currently if `metabrowse` is enabled, attempting to get contextual help of a symbol (shift-tab) would result in a kernel restart due to failure to start the metabrowse server. Happens both with Scala 2.12.10 and 2.13.1.

```
java.lang.VerifyError: Bad return type
Exception Details:
  Location:
    scala/meta/internal/mtags/MtagsEnrichments.scala$meta$internal$mtags$MtagsEnrichments$$filenameToLanguage(Ljava/lang/String;)Lscala/meta/internal/semanticdb/Language; @33: ar
eturn
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to 'scala/meta/internal/semanticdb/Language' (from method signature)
  Current Frame:
    bci: @33
    flags: { }
    locals: { 'scala/meta/internal/mtags/MtagsEnrichments', 'java/lang/String' }
    stack: { 'java/lang/Object' }
  Bytecode:
    0000000: 2b12 74b6 007a 9900 09b2 007e a700 152b
    0000010: 1280 b600 7a99 0009 b200 83a7 0006 b200
    0000020: 86b0
  Stackmap Table:
    same_frame(@15)
    same_frame(@30)
    same_locals_1_stack_item_frame(@33,Object[#4])

        at scala.meta.internal.mtags.OnDemandSymbolIndex.$anonfun$addSourceJar$2(OnDemandSymbolIndex.scala:56)
        at scala.meta.internal.mtags.OnDemandSymbolIndex.$anonfun$addSourceJar$2$adapted(OnDemandSymbolIndex.scala:55)
        at scala.meta.internal.io.PlatformFileIO$.withJarFileSystem(PlatformFileIO.scala:87)
        at scala.meta.internal.io.FileIO$.withJarFileSystem(FileIO.scala:44)
        at scala.meta.internal.mtags.OnDemandSymbolIndex.$anonfun$addSourceJar$1(OnDemandSymbolIndex.scala:55)
        at scala.meta.internal.mtags.OnDemandSymbolIndex.tryRun(OnDemandSymbolIndex.scala:107)
        at scala.meta.internal.mtags.OnDemandSymbolIndex.addSourceJar(OnDemandSymbolIndex.scala:54)
        at metabrowse.server.MetabrowseServer.$anonfun$replaceClasspath$2(MetabrowseServer.scala:79)
        at metabrowse.server.MetabrowseServer.$anonfun$replaceClasspath$2$adapted(MetabrowseServer.scala:79)
        at scala.collection.immutable.List.foreach(List.scala:392)
        at metabrowse.server.MetabrowseServer.replaceClasspath(MetabrowseServer.scala:79)
        at metabrowse.server.MetabrowseServer.start(MetabrowseServer.scala:48)
        at almond.internals.ScalaInterpreterInspections.createMetabrowseServer(ScalaInterpreterInspections.scala:105)
        at almond.internals.ScalaInterpreterInspections.$anonfun$metabrowseServerOpt$2(ScalaInterpreterInspections.scala:40)
        at scala.Option.orElse(Option.scala:447)
        at almond.internals.ScalaInterpreterInspections.$anonfun$metabrowseServerOpt$1(ScalaInterpreterInspections.scala:39)
        at scala.Option.orElse(Option.scala:447)
        at almond.internals.ScalaInterpreterInspections.metabrowseServerOpt(ScalaInterpreterInspections.scala:38)
        at almond.internals.ScalaInterpreterInspections.inspect(ScalaInterpreterInspections.scala:111)
        at almond.ScalaInterpreter.inspect(ScalaInterpreter.scala:151)
        at almond.interpreter.Interpreter.inspect(Interpreter.scala:146)
        at almond.interpreter.Interpreter.inspect$(Interpreter.scala:145)
        at almond.ScalaInterpreter.inspect(ScalaInterpreter.scala:21)
        at almond.interpreter.util.AsyncInterpreterOps.asyncInspect(AsyncInterpreterOps.scala:21)
        at almond.interpreter.util.AsyncInterpreterOps.asyncInspect$(AsyncInterpreterOps.scala:20)
        at almond.ScalaInterpreter.asyncInspect(ScalaInterpreter.scala:21)
        at almond.ScalaInterpreter.asyncInspect(ScalaInterpreter.scala:21)
        at almond.interpreter.InterpreterToIOInterpreter.$anonfun$inspectionCancellable$4(InterpreterToIOInterpreter.scala:139)
        at almond.interpreter.util.Cancellable.run(Cancellable.scala:32)
        at almond.interpreter.InterpreterToIOInterpreter.inspect(InterpreterToIOInterpreter.scala:150)
        at almond.interpreter.messagehandlers.InterpreterMessageHandlers.$anonfun$inspectHandler$1(InterpreterMessageHandlers.scala:147)
        at almond.interpreter.messagehandlers.MessageHandler$.$anonfun$blocking$2(MessageHandler.scala:127)
        at almond.interpreter.messagehandlers.MessageHandler$.$anonfun$blockingTaskStream$1(MessageHandler.scala:157)
        at cats.effect.IO$Map.apply(IO.scala:1699)
        at cats.effect.IO$Map.apply(IO.scala:1697)
        at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:187)
        at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:463)
        at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:484)
        at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:422)
        at cats.effect.internals.IOShift$Tick.run(IOShift.scala:36)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base/java.lang.Thread.run(Thread.java:832)
```

### PR DESCRIPTION
**Pull request by kiendang**
Close #764
Close #183

This is a POC PR for displaying type info and docstrings upon symbol inspection without metabrowse by reusing the implementation from the LSP. This depends on `mtags` and copies over the relevant code from `metals` with some minor adjustments.

### Demo

Can display type info and docstrings in Jupyter contextual help upon inspection (Shift-Tab)

type
> almond_inspect_type_docstrings

method
> almond_inspect_method_docstrings

def
> almond_inspect_def_def

val
> almond_inspect_val

### Things that do not work (yet)

- Can't display info for builtin objects like `kernel`, `repl`. I guess we can do this by exposing and adding `repl.fullImports` to `code`.
- Can't display docstrings defined in the current session. Might be able to do this by adding a separate docstrings search implementation for things in the current file.

> almond_inspect_now_defined_docstrings

### Discuss

This definitely needs polishing and is not ready to be merged yet. Just wanna find out if this is the right direction.

### CODE REVIEW

**Review by alexarchambault — COMMENTED** [review]
Thanks @kiendang! That's really neat!

I didn't review that in detail yet. But do you think this could be added alongside the current metabrowse support? Either enabling it for Scala versions that lack metabrowse support, or enabling it all the time, but keep the metabrowse server and link for Scala versions that allow it. I'm still hoping to update metabrowse myself at some point, and its features are a nice complement to this.

Also, are all copy-pasted classes from mtags necessary? It seems `MetalsSymbolDocumentation`, `JavadocIndexer`, and `EmptySymbolDocumentation` could be re-used from mtags, rather than being copied here. Isn't it?

**Comment by kiendang** [comment]
Thanks for reviewing!

> do you think this could be added alongside the current metabrowse support?

Yup I'll put that back in

> Also, are all copy-pasted classes from mtags necessary? It seems `MetalsSymbolDocumentation`, `JavadocIndexer`, and `EmptySymbolDocumentation` could be re-used from mtags, rather than being copied here. Isn't it?

I didn't copy paste from mtags, only from metals. So `Docstrings`, `ScaladocIndexer`, `JavadocIndexer`, `MetalsSymbolDocumentation` and `EmptySymbolDocumentation` are not in mtags but in metals. I already add mtags as a dependency.
Personally I would love to see `Docstrings`, `ScaladocIndexer` and `JavadocIndexer`got moved from metals to mtags too. Looking up docstrings from symbol seems like a useful task that sees usage outside of metals (like here) and would ease the maintenance on almond's side. And according to https://github.com/almond-sh/almond/issues/183#issuecomment-426916796 looks like previously it was the original idea?
However I'm not familiar with metals repo. Would it be a good idea to raise an issue there that linked to this?

**Comment by alexarchambault** [comment]
Precisely, I think `org.scalameta:::mtags` lives [in the metals repository](https://github.com/scalameta/metals/tree/23c25680c5d4b1e1114ac425efa9e5bcc3eafb39/mtags/src/main). So it should be around.

**Comment by kiendang** [comment]
So currently metabrowse depends on mtags 0.8.4 while this PR uses 0.10.0. I have a problem with mtags 0.8.4, the same one that causes metabrowse to not work in #764. Could you bump mtags in metabrowse 0.8.4 -> 0.10.0?

**Comment by tgodzik** [comment]
The only rather stable interface in Metals is `PresentationCompiler` java interface, would you be able to reuse that to get what you need? We have never actually thought of mtags as a library the people could use, so maybe we can add some additional API and make sure it's MIMAd ? I've seen a couple of instances that this might have been useful especially if we don't want to start the entire Metals serwer.

Alternatively (just shouting out an idea), there were some plans to reuse Almond in Metals itself using the VS Code notebook API, but I am just a bit lost in the this area, so not sure if it would overall help the ecosystem.

TLDR; We can expose a stable API in Metals if needed, we just haven't really thought of it. Anything internal is not safe currently to use.

**Comment by kiendang** [comment]
Thanks for taking a look at this!

> The only rather stable interface in Metals is `PresentationCompiler` java interface, would you be able to reuse that to get what you need?

I ported over the `documentation` method in `StandaloneSymbolSearch` here so I guess it's the `SymbolSearch`
 interface?

> We can expose a stable API in Metals if needed

That would be great. If I'm not wrong originally there were talks about exposing an API to go from symbol to docstrings https://github.com/almond-sh/almond/issues/183#issuecomment-426931034

> Alternatively (just shouting out an idea), there were some plans to reuse Almond in Metals itself using the VS Code notebook API, but I am just a bit lost in the this area, so not sure if it would overall help the ecosystem.

I've been using notebooks in VS Code for Python and really like it. Would be nice to have Almond there too. But AFAIK supports for language kernels other than Python in VS Code Notebook API is not mature yet. Like you could write an extension that makes Almond works but it will conflict with the Python one. But as a user this is something I would like to see happen eventually.

**Comment by tgodzik** [comment]
> I ported over the documentation method in StandaloneSymbolSearch here so I guess it's the SymbolSearch
interface?

From what I see `GlobalSymbolIndex` is only used here in the PR? 

Anyway, what I think we might do:
- open an issue in Metals explaining what we don't to expose out of mtags/metals
- move Scala interface from internal to `scala.meta` with stable methods
- we keep that one stable (though mtags do not adhere to any binary compat, which might be an issue)

Alternatively it might be better to copy over the code if it's small enough.

I would start with an issue in Metals, where all interested parties can discuss what interfaces they would need. Otherwise, we will forget about this the same as with the previous case :sweat:

**Comment by kiendang** [comment]
Sure. I'll prepare a minimal snippet to make it easier to see what's needed to expose and open an issue over at the metals repo.

**Comment by alexarchambault** [comment]
Now part of https://github.com/almond-sh/almond/pull/1110
